### PR TITLE
Dupuit

### PR DESCRIPTION
**Pull request by TomFischer**
Implementation of Dupuit permeability relation.

 - [x] Changelog entry
 - [x] documentation
 - [x] tests / benchmarks

**Update**
Later PR:
 - storage model

### CODE REVIEW

**Comment by TomFischer** [comment]
2019-05-14: rebased

**Comment by Thomas-TK** [comment]
@TomFischer 
- documentation and 4 benchmark tests are prepared. 
We would need now a place to upload them. I would call the headline "Dupuit assumption" or "unconfined groundwater flow".

**Comment by Thomas-TK** [comment]
Just wondering while looking at the benchmark documentation: if we have a headline called elliptic, why don't we have parabolic as well?

**Comment by Thomas-TK** [comment]
. . I guess I found the answer: web\content\docs\benchmarks. To be prepared with markdown. The web-folder contains a readme :-)

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2018, OpenGeoSys Community (http://www.opengeosys.org)
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]:
```suggestion
 * Copyright (c) 2012-2019, OpenGeoSys Community (http://www.opengeosys.org)
```

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+{
+namespace PorousMedium
+{
+/// The class implements a basic permeability model that employs a parameter
+/// (for instance a constant parameter or mesh cell dependend parameter) to fill
+/// the intrinsic permeability tensor.
+class DupuitPermeability final : public Permeability
+{
+public:
+    explicit DupuitPermeability(
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]:
```suggestion
    DupuitPermeability(
```

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        ParameterLib::Parameter<double> const& permeability_parameter,
+        int const dimension)
+        : Permeability(permeability_parameter, dimension)
+    {
+    }
+
+    ~DupuitPermeability() = default;
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]:
```suggestion
```

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+     *  @param variable    A variable with any double type value.
+     *  @param temperature Temperature with any double type value.
+     */
+    Eigen::MatrixXd const& getValue(
+        const double t,
+        ParameterLib::SpatialPosition const& pos,
+        const double variable,
+        const double temperature) const override
+    {
+        _intrinsic_permeability_tensor =
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: Why storing a value which is recomputed every time?
```suggestion
        return
```

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        ParameterLib::Parameter<double> const& permeability_parameter,
+        int const dimension)
+        : Permeability(permeability_parameter, dimension)
+    {
+    }
+
+    ~DupuitPermeability() = default;
+
+    /**
+     *  Get property value.
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: It's not property value, it's the permeability tensor. If possible, do not duplicate docu from the base class; doxygen should be doing this.

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        const double temperature) const override
+    {
+        _intrinsic_permeability_tensor =
+            variable * Permeability::getValue(t, pos, variable, temperature);
+
+        return _intrinsic_permeability_tensor;
+    }
+
+private:
+//    ProcessLib::Parameter<double> const& _thickness;
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]:
```suggestion
```

File: ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h [hunk 7]
```diff
@@ -33,7 +33,7 @@ void LiquidFlowLocalAssembler<ShapeFunction, IntegrationMethod, GlobalDim>::
     const int material_id = _material_properties.getMaterialID(pos);
 
     const Eigen::MatrixXd& permeability = _material_properties.getPermeability(
-        material_id, t, pos, _element.getDimension());
+        material_id, t, pos, _element.getDimension(), 0.0, 0.0);
```
Comment (endJunction) [inline ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h]: what are the magic numbers?

File: web/content/docs/benchmarks/liquid-flow/unconfined-aquifer.pandoc [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+- The density gets the value 1:			ρ=1
+- The viscosity gets the value 1:		μ=1
+
+Note: in such a case, the result is also an output of hydraulic head in [m] and not Pressure in [Pa].
+
+
+
+## Examples:
+
+The folowing simple examples, which have been compared with Modflow simulations, shall verify the result and demonstrate the basic parameterization.
```
Comment (endJunction) [inline web/content/docs/benchmarks/liquid-flow/unconfined-aquifer.pandoc]:
```suggestion
The following simple examples, which have been compared with Modflow simulations, shall verify the result and demonstrate the basic parameterization.
```

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h (hunk 1 above)
Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: fixed

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h (hunk 2 above)
Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: fixed

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h (hunk 3 above)
Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: fixed

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h (hunk 5 above)
Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: Referenced the (updated) base class documentation.

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h (hunk 6 above)
Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: removed

File: ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h (hunk 7 above)
Comment (TomFischer) [reply ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h]: Substituted magic numbers by constants initialized with nans.

File: web/content/docs/benchmarks/liquid-flow/unconfined-aquifer.pandoc (hunk 8 above)
Comment (TomFischer) [reply web/content/docs/benchmarks/liquid-flow/unconfined-aquifer.pandoc]: fixed

File: ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h [hunk 9]
```diff
@@ -104,6 +105,10 @@ void LiquidFlowLocalAssembler<ShapeFunction, IntegrationMethod, GlobalDim>::
         const double mu =
             _material_properties.getViscosity(p, _reference_temperature);
 
+        pos.setIntegrationPoint(ip);
+        auto const& permeability = _material_properties.getPermeability(
+            material_id, t, pos, _element.getDimension(), p, 0.0);
```
Comment (endJunction) [inline ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h]:
```suggestion
            material_id, t, pos, _element.getDimension(), p, _reference_temperature);
```
It's used above in `mu` computation.

File: ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h [hunk 10]
```diff
@@ -182,6 +191,8 @@ void LiquidFlowLocalAssembler<ShapeFunction, IntegrationMethod, GlobalDim>::
         const double mu =
             _material_properties.getViscosity(p, _reference_temperature);
 
+        auto const& permeability = _material_properties.getPermeability(
+            material_id, t, pos, _element.getDimension(), p, 0.0);
```
Comment (endJunction) [inline ProcessLib/LiquidFlow/LiquidFlowLocalAssembler-impl.h]: dto.

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h (hunk 4 above)
Comment (endJunction) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: ?

File: MaterialLib/PorousMedium/Permeability/Permeability.h [hunk 11]
```diff
@@ -68,7 +69,7 @@ class Permeability
         return _intrinsic_permeability_tensor;
     }
 
-private:
+protected:
     ParameterLib::Parameter<double> const& _permeability_parameter;
     int const _dimension;
     mutable Eigen::MatrixXd _intrinsic_permeability_tensor;
```
Comment (endJunction) [inline MaterialLib/PorousMedium/Permeability/Permeability.h]: Try to remove this member. I don't see where it is used. In two `getValue` functions the tensor is computed every time.

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h (hunk 4 above)
Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: Tried the suggested change. It crashes the benchmarks. Don't know why.

Comment (TomFischer) [reply MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: Removing `_intrinsic_permeability_tensor` leads to the compiler warning:
```
 Permeability.h:60:16: warning: returning reference to local temporary object [-Wreturn-stack-address]
        return Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
```

**Comment by wenqing** [comment]
One can set mu=1, rho=0 in the LiquidFlow to use pressure as head. Then one has to keep in mind that the permeability variable keeps the value of hydraulic conductivity. This would cause some sort of confusion.

**Comment by endJunction** [comment]
@wenqing Would this also be applicable to HT, for example? As I understood it, the "permeability model" can be used in other processes too.

**Comment by wenqing** [comment]
@endJunction I think it is only for a groundwater model. As far as I could understand that this PR wants to handle hK in \beta h = grad (h K grad h) for large spatial groundwater problem by introducing a permeability model of hK.  You can see `h K` cannot be considered as  hydraulic conductivity. The problem is just how to deal with unknowns. Furthermore the nonlinearity of grad (h K grad h) has to be dealt properly (better use Newton-Raphson) for the solution accuracy.

**Comment by wenqing** [comment]
I have checked that the implementation of the unconfined equation is correct. As discussed that  there should be explanation in the documentation of DupuitPermeability that DupuitPermeability provides h *K for unconfined flow equation, and its name is meaningless  for such case.

**Comment by TomFischer** [comment]
@wenqing Please, could you check the changed documentation of DupuitPermeability class.

File: MaterialLib/PorousMedium/Permeability/DupuitPermeability.h [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+#include "MaterialLib/PorousMedium/Permeability/Permeability.h"
+#include "ParameterLib/Parameter.h"
+
+namespace MaterialLib
+{
+namespace PorousMedium
+{
+/// The purpose of the class DupuitPermeability is the implementation of the
+/// special diffusion coefficient \f$h K(h)\f$, where \f$h\f$ is the hydraulic
+/// head and \f$K\f$ is the hydraulic conductivity. The diffusion coefficient is
```
Comment (wenqing) [inline MaterialLib/PorousMedium/Permeability/DupuitPermeability.h]: `... the hydraulic head and ...` -->`... the hydraulic head of the previous non-linear iteration and ...`

**Pull request merged**